jplag/src/main/java/de/jplag/clustering/algorithm/BayesianOptimization.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.commons.math3.exception.OutOfRangeException;
+import org.apache.commons.math3.linear.ArrayRealVector;
+import org.apache.commons.math3.linear.RealVector;
+import org.apache.commons.math3.random.HaltonSequenceGenerator;
+import org.apache.commons.math3.random.RandomVectorGenerator;
+
+/**
+ * Maximizes a function using bayesian optimization.
+ */
+public class BayesianOptimization {
+
+    private static final int STOP_AFTER_CONSECUTIVE_RANDOM_PICKS = 3;
+    private static final int MAX_NON_ZERO_ACQ_FN_EVALS_PER_ITERATION = 50;
+    private static final int MAXIMUM_ACQ_FN_EVALS_PER_ITERATION = 1000;
+
+    private RealVector minima;
+    private RealVector maxima;
+    private int maxEvaluations;
+    private int initialPoints;
+    private double noise;
+    private RealVector lengthScale;
+    private boolean debug = false;
+
+    /**
+     * @param minima of the explored parameters
+     * @param maxima of the explored parameters
+     * @param initPoints points that are initially sampled for exploration
+     * @param maxEvaluations maximal evaluations of the fitted function
+     * @param noise of the explored function
+     * @param lengthScale width parameter for the matern kernel
+     */
+    public BayesianOptimization(RealVector minima, RealVector maxima, int initPoints, int maxEvaluations, double noise, RealVector lengthScale) {
+        if (minima.getDimension() != maxima.getDimension()) {
+            throw new DimensionMismatchException(minima.getDimension(), maxima.getDimension());
+        }
+        if (initPoints < 1 || initPoints > maxEvaluations) {
+            throw new OutOfRangeException(initPoints, 1, maxEvaluations);
+        }
+        this.maxima = maxima;
+        this.minima = minima;
+        this.initialPoints = initPoints;
+        this.maxEvaluations = maxEvaluations;
+        this.noise = noise;
+        this.lengthScale = lengthScale;
+    }
+
+    public void setDebugLogging(boolean debug) {
+        this.debug = debug;
+    }
+
+    private Stream<RealVector> sampleSolutionSpace() {
+        RandomVectorGenerator generator = new HaltonSequenceGenerator(minima.getDimension());
+        RealVector size = maxima.subtract(minima);
+        return Stream.generate(generator::nextVector).map(x -> new ArrayRealVector(x)).map(x -> x.ebeMultiply(size)).map(x -> x.add(minima));
+
+    }
+
+    private GaussianProcess fit(List<RealVector> X, List<Double> Y) {
+        return GaussianProcess.fit(X, Y.stream().mapToDouble(Double::doubleValue).toArray(), noise, true, lengthScale.toArray());
+    }
+
+    private double acquisitionFunction(GaussianProcess gpr, double[] r, double yMax) {
+        // expected improvement
+        double[] meanAndStandardDeviation = gpr.predictWidthStd(new ArrayRealVector(r));
+        double mean = meanAndStandardDeviation[0];
+        double standardDeviation = meanAndStandardDeviation[1];
+        double a = mean - yMax;
+        double z = a / standardDeviation;
```

Comment:
You know the drill! ;)

Reply:
Tbh I don't even know how to call half of the intermediate values of the algorithms I implemented.